### ISSUE TITLE
OIIO python api is not properly overwriting an image

### ISSUE THREAD

**Issue by dekekincaid** [OpenImageIO/oiio#1987]
### Problem

We are running into an issue where we are writing an image "A" with an external DCC and then using oiio to manipulate the file and rewrite over top of the same image with the same file name which I will call image "B".  When we view the file with iv, I see image A.

**Expected behavior:**  
Image B should be in place, not A.

**Actual behavior:**  
I see image A.

### Steps to Reproduce
Make two jpg images(checkerboard) with different sizes (tested with both jpg & exr).
dog.jpg = 640x480
cow.jpg = 1280x720

This will make it easy to verify whether the OIIO python API has the correct image without opening the image in a viewer.

We will use an oiio.ImageBuf to repeatedly open a file called animal.jpg which is created by copying either dog.jpg or cow.jpg to a file with the name animal.jpg

# python code below:
```

import OpenImageIO as oiio

# before proceeding
# in an external shell, copy dog.jpg to animal.jpg
filepath_in = "/var/tmp/animal.jpg"

image = oiio.ImageBuf(filepath_in)
image.read()
print image.spec().width

# should print 640 to the console here

# wipe the contents of the ImageBuf
image.clear()

# now, in an external shell, copy cow.jpg over top of animal.jpg

image = oiio.ImageBuf(filepath_in)

image.read()
print image.spec().width

# should print 1280 to the console here, but prints 640 instead.

# end of code
```

You'll notice that not only are the width and height of the original incarnation of animal.jpg frozen for eternity, the image content is as well. In addition to ImageBuf.clear() we've also tried reset() and read(0,0,True) to force the flushing of the cache. But OIIO python API never seems to be able to dissociate the file name animal.jpg with anything but its first incarnation. Is this a bug? Or is there a documented way to re-read a file whose content has changed. Thanks in advance.

### Versions

* OIIO 1.8.5 (also tested with 1.8.10)
* OS: Centos 6.4
* C++ compiler: gcc 4.8.5

**Comment by lgritz** [comment]
What you need to understand is that by default, ImageBuf reads are done lazily, and are backed by ImageCache. So the sequence you've laid out is:

- declare ImageBuf for animal.jpg (lazy -- won't even open the file yet)
- `image.read()` still lazy -- header is read now, but as an aside, pixels still aren't read
- delete the buffer (note: header info and any pixels you accessed are still in ImageCache)
- unbeknownst to the ImageCache, the image is changed on disk
- another ImageBuf names animal.jpg
- read the buffer, and it gets the info that's still in the ImageCache from earlier

There are two ways to address this.

1. If you don't want the ImageCache involved at all, and you want to be sure that the read() always does a complete read of whatever is on disk at that moment, you want:

        image.read (force=True)

2. If you still want ImageCache to do its magic (and by the way, in this case, the `read()` call itself is unnecessary, it will automatically read when you access header info or pixels), and you know when an external program might change the file, then you can do this:

        image.clear ()
        oiio.ImageCache.create().invalidate (filename_to_invalidate)
        // ... now the next ImageBuf to reference that file will start anew

**Comment by lgritz** [comment]
Did this explanation help?

**Comment by dekekincaid** [comment]
Yes, sorry, I meant to reply to this earlier.  We found another thread on oiio-dev which you kind of explained the same thing from earlier in the year.  It would be good to have this in the oiio python docs though.  This isn’t explained in there.

Cheers

**Comment by lgritz** [comment]
I think I have a technical fix that should help prevent people running into this. Coming soon.

**Comment by filmkorn** [comment]
> oiio.ImageCache.create().invalidate (filename_to_invalidate)

@lgritz We have this line in a few places in our code. As this is no longer compatible with oiio-2 I was wondering if you implemented the technical fix you were thinking of?

**Comment by lgritz** [comment]
I think ImageCache.create() is removed because it's unnecessary. Just use the constructor:

    oiio.ImageCache().invalidate (filename)

Does that work?

Now, as for the question about whether I implemented the technical fix... uh, I don't know, this was a while ago and I need to see what I did about it. Let me poke around a bit in my records and look through closed PRs, maybe I just never associated it with this issue.

**Comment by lgritz** [comment]
I'm re-opening this ticket. I think the fact that it was closed confused me before into thinking it was resolved, but I think I never added the feature I was hypothesizing would make this trick unnecessary.

**Comment by lgritz** [comment]
Proposed automatic fix for this is in PR #2696

**Comment by lgritz** [comment]
I've merged the automatic fix into master. I want to let it sit there for a little while to let it get more extensively tested, but if no problem is found with it, I will backport it to 2.2 in time for the Oct 1 monthly patch release.
